Sources/firebase/include/FirebaseCore.hh

```diff
@@ -1,7 +1,34 @@
-
 #ifndef firebase_include_FirebaseCore_hh
 #define firebase_include_FirebaseCore_hh
 
 #include <firebase/util.h>
 
+#include <stdio.h>
+
+namespace swift_firebase::swift_cxx_shims::firebase {
+
+typedef void (*FutureCompletionType)(void*);
+
+// This class exists to provide protocol conformance to FutureProtocol.  It
+// also provides a method to invoke `OnCompletion` in a way that works from
+// Swift. We can ignore the `FutureBase` param as the Swift caller can just
+// retain the Future as part of its closure.
+template <class R> class ConformingFuture: public ::firebase::Future<R> {
+ public:
+  typedef R ResultType;
+  typedef ::firebase::Future<R> FutureType;
+
+  ConformingFuture(const FutureType& rhs) : FutureType(rhs) {}
+
+  void CallOnCompletion(
+      _Nonnull FutureCompletionType completion,
+      _Nullable void* user_data) const {
+    OnCompletion([completion, user_data](const FutureBase&) {
+      completion(user_data);
+    });
+  }
+} __attribute__((swift_attr("conforms_to:FirebaseCore.FutureProtocol")));
```

Comment:
Please move the attribute to after the `class` keyword.

Reply:
we typically recommend putting it after the class definition (after the `}` like you have here) as per the docs, but there's not a strong rule here.  My preference would be to suggest leaving it here where you have it now

Reply:
IIRC, C++ recommends that the attributes follow the declarator (`class`). This is where `__declspec` and `__attribute__` are immediately parsed. My preference would be to follow C++ parser behaviour rather than GNU extensions.

Reply:
I don't have a strong opinion, so happy to switch to match your preference. I was just following the guidance from the Swift C++ documentation that @hyp pointed me to. Seemed reasonable.

Reply:
Moved